src/fhirtypes/ElementDefinition.ts

```diff
@@ -972,6 +972,22 @@ export class ElementDefinition {
     const targetType = this.getTargetType(target, fisher);
     const targetTypes: ElementDefinitionType[] = targetType ? [targetType] : this.type;
 
+    // Check if a CodeableReference is attempting to constraint directly to the reference element
+    if (
+      this.type.length === 1 &&
+      this.type[0].code === 'Reference' &&
+      this.path.endsWith('.reference') &&
+      this.parent()?.type?.[0]?.code === 'CodeableReference'
+    ) {
+      const errorMessage =
+        'Constraining references on CodeableReference elements should not be applied directly on the .reference element. ' +
+        `The constraint should be applied to the parent element: ${this.path
+          .split('.')
+          .slice(1, -1) // slice of the resource name and the final .concept portion in order to match the recommended FSH path
+          .join('.')}`;
+      logger.error(errorMessage);
```

Comment:
I think it would be useful to include `rule.sourceInfo` in the logged message. Also, if the rule was applied with a path that includes a slice, this message won't include it, since `this.path` doesn't include slices. So, maybe it should use `this.id` instead?

Reply:
Great point on both. I wonder if maybe trying to include the path is too much here. Could there be places where the `this.id` value doesn't actually match what the FSH path should be? Like for slices, won't the syntax not match? I think maybe this message is trying to be too helpful and is becoming unhelpful. I'm leaning towards just taking out the second sentence and leaving the first. This would apply to both the error messages.

Reply:
I sort of just went for it and removed the path sentence because I talked myself into thinking this was adding unnecessary complication, especially now that the rule info is listed and the line number will be right there. So [e2d9bdb](https://github.com/FHIR/sushi/pull/1294/commits/e2d9bdbc5b443fd198efd3a280049be84b6bdcc3) removes that second sentence where the path could have been wrong.

That same commit also adds the rule source info to the message. I did have to pass in `rule.sourceInfo` to `bindToVs` in order to support that. I made it an optional parameter in the function because there are some places where we bind without it coming directly from a rule.